src/rpc/methods/eth/pubsub.rs

```diff
@@ -59,172 +59,89 @@
 //! ```
 //!
 
-use std::fmt;
-
+use crate::rpc::eth::pubsub_trait::{
+    EthPubSubApiServer, LogFilter, SubscriptionKind, SubscriptionParams,
+};
+use crate::rpc::{RPCState, chain};
 use fvm_ipld_blockstore::Blockstore;
-use serde::de::{self, Deserializer, SeqAccess, Visitor};
-use serde::{Deserialize, Serialize};
+use jsonrpsee::PendingSubscriptionSink;
+use jsonrpsee::core::{SubscriptionError, SubscriptionResult};
+use std::sync::Arc;
 use tokio::sync::broadcast::{Receiver as Subscriber, error::RecvError};
 
-use crate::rpc::Ctx;
-use crate::rpc::eth::types::EthAddressList;
-use crate::rpc::eth::{EthFilterSpec, EthTopicSpec};
-
-pub const ETH_SUBSCRIPTION: &str = "eth_subscription";
-
-const NEW_HEADS: &str = "newHeads";
-const PENDING_TRANSACTIONS: &str = "pendingTransactions";
-const LOGS: &str = "logs";
-
-#[derive(Default, Serialize, Deserialize, Debug, Clone)]
-#[serde(rename_all = "camelCase")]
-pub struct LogFilter {
-    pub address: EthAddressList,
-    pub topics: Option<EthTopicSpec>,
+pub struct EthPubSub<DB> {
+    ctx: Arc<RPCState<DB>>,
 }
 
-#[derive(Debug)]
-enum Subscription {
-    NewHeads,
-    PendingTransactions,
-    Logs(Option<LogFilter>),
+impl<DB> EthPubSub<DB> {
+    pub fn new(ctx: Arc<RPCState<DB>>) -> Self {
+        Self { ctx }
+    }
 }
 
-impl<'de> Deserialize<'de> for Subscription {
-    fn deserialize<D>(deserializer: D) -> Result<Subscription, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        struct SubscriptionVisitor;
-
-        impl<'de> Visitor<'de> for SubscriptionVisitor {
-            type Value = Subscription;
-
-            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-                formatter.write_str(r#"a JSON array like ["logs", {...}] or ["newHeads"]"#)
+#[async_trait::async_trait]
+impl<DB> EthPubSubApiServer for EthPubSub<DB>
+where
+    DB: Blockstore + Send + Sync + 'static,
+{
+    async fn subscribe(
+        &self,
+        pending: PendingSubscriptionSink,
+        kind: SubscriptionKind,
+        params: Option<SubscriptionParams>,
+    ) -> SubscriptionResult {
+        let sink = pending.accept().await?;
+        let ctx = self.ctx.clone();
+
+        match kind {
+            SubscriptionKind::NewHeads => self.handle_new_heads_subscription(sink, ctx).await,
+            SubscriptionKind::PendingTransactions => {
+                // TODO(akaladarshi): https://github.com/ChainSafe/forest/pull/5782
+                return Err(SubscriptionError::from(
+                    jsonrpsee::types::ErrorObjectOwned::owned(
+                        jsonrpsee::types::error::METHOD_NOT_FOUND_CODE,
+                        "pendingTransactions subscription not yet implemented",
+                        None::<()>,
+                    ),
+                ));
```

Comment:
Does it reference the right issue? Seems to be referencing the same PR.

Reply:
Removing the reference from here, since there is no specific issue for it.